### DISCUSSION TITLE
getter 함수에서 DeepCopy를 수행하는 것에 대해 어떻게 생각하시나요?

### DISCUSSION THREAD

**Discussion by wonsss** [woowacourse-study/2022-code-review-study#1] [Lotto] [answered]
- [#81] Model의 getter 함수에서 DeepCopy를 수행하는 것에 대해 어떻게 생각하시나요 ?
  - 질문자의 질문의도: get으로 가져오는 멤버 변수는 불러온 사용처에서 변경을 가하더라도 영향이 가지 않았으면 해요. 사용된 곳에서 가한 변경이 객체의 멤버에 직접 전달되어 버린다면, 예상치 못한 부수효과들이 발생할 것 같거든요 하지만 DeepCopy 하는 로직의 경우 객체의 레이어가 깊을수록, 데이터가 많을수록 비용이 너무 커질 것 같아 우려되기도 합니다.

```jsx
// 예시
// 유틸
export const deepCopy = object => JSON.parse(JSON.stringify(object));

// 모델
class LottoGameModel {
  constructor() {
    this.lottoList = null;
  }

  getLottoList() {
    /** getter로 가져간 lottoList를 변경하여도 lottoList의 멤버에겐 영향이 없다. */
    return deepCopy(this.lottoList);
  }
}

**Comment by wonsss** [comment, marked as answer]
[Private class fields](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Classes/Private_class_fields)를 사용하고, 메서드로 private [접근자](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Functions/get)를 사용하면, 원본의 불변성도 보장되는 줄 알았는데, 그게 아니었네요.
 Private class fields를 쓴 멤버변수에 대해 get 접근자를 써야, 접근하여 값을 받을 수 있죠. 하지만 반환받은 그 값이 객체일 경우 객체의 프로퍼티를 조작하면, 원본인 인스턴스의 데이터도 변경되는 side effect가 발생해요.
따라서, 이러한 side effect가 일어날 수 있는 상황이라고 예상한다면, private class fields라 하더라도 깊은 복사를 하여 반환하는 방법을 고려해볼 필요가 있는 것 같아요.

**Reply by lokba** [reply]
마르코, 혹시 참고한 블로그 같은게 있을까요? 
궁금하네요..

**Reply by wonsss** [reply]
저도 검색해봤는데 관련된 내용 찾기가 어렵더라구요. 그래서 직접 코드 만들어서 테스트해봤어요.ㅠㅠ

**Reply by wonsss** [reply]
- 객체의 값이 있는 private 필드를 return 받아 변수에 할당하고, 객체인 그 변수의 프로퍼티 값을 변경한 결과, 원본(인스턴스의 private 프로퍼티)도 변경되는 예시입니다.
```jsx
class test {
  #privateObject;

  constructor() {
    this.#privateObject = {
      deep: { deeper: { moreDeeper: "marco" } },
      age: 100,
    };
  }

  get privateObject() {
    return this.#privateObject;
  }
}

const testInstance = new test();
const targetObject = testInstance.privateObject;
targetObject.age = 20;
targetObject.deep.deeper.moreDeeper = "lokba";
console.log(targetObject); // { deep: { deeper: { moreDeeper: 'lokba' } }, age: 20 }
console.log(testInstance.privateObject);  // { deep: { deeper: { moreDeeper: 'lokba' } }, age: 20 }
```

**Reply by wonsss** [reply]
- getter에서 반환 전에 깊은 복사를 하여 반환한 경우에는, side effect를 방지할 수 있는 것 같군용. 그런데 이와 관련한 단점 또는 over spec일 수 있다는 점도 고민해보면 좋을 것 같아요. 테스트코드도 붙입니다.
```jsx
const deepCopy = object => JSON.parse(JSON.stringify(object));

class test {
  #privateObject;

  constructor() {
    this.#privateObject = {
      deep: { deeper: { moreDeeper: "marco" } },
      age: 100,
    };
  }

  get privateObject() {
    return deepCopy(this.#privateObject);
  }
}

const testInstance = new test();
const targetObject = testInstance.privateObject;
targetObject.age = 20;
targetObject.deep.deeper.moreDeeper = "lokba";
console.log(targetObject); // { deep: { deeper: { moreDeeper: 'lokba' } }, age: 20 }
console.log(testInstance.privateObject);  // { deep: { deeper: { moreDeeper: 'marco' } }, age: 100 }
```

**Reply by lokba** [reply]
너무 고생하셨어요. 따봉 3개입니다 마르코~  👍 👍 👍 

해당 내용은 블로그 포스팅을 하면 더 좋을 것 같기도 하네요. 마르코 관심있으면, 블로그 포스팅하고 크루원들에게 공유해도 정말 좋을것 같아요.
마르코 덕분에 새롭게 알게된 정보를 얻게 되네요. 

내일 단점에 대해서도 한번 생각해볼게요

**Reply by moonheekim0118** [reply]
아하~ 객체의 불변성을 지켜주시고싶다는 말씀이시군요..좋은 포인트네요. getter에서도 불변성을 지켜야하는걸 깨달았네요..
개인적으로 deepCopy 해주는게 안전해서 좋다고 생각합니다. 굿굿
